File: android/src/main/kotlin/twilio/flutter/twilio_programmable_chat/listeners/ChannelListener.kt
package twilio.flutter.twilio_programmable_chat.listeners

import com.twilio.chat.Channel
import com.twilio.chat.ChannelListener as TwilioChannelListener
import com.twilio.chat.ErrorInfo
import com.twilio.chat.Member
import com.twilio.chat.Message
import io.flutter.plugin.common.EventChannel
import twilio.flutter.twilio_programmable_chat.Mapper
import twilio.flutter.twilio_programmable_chat.TwilioProgrammableChatPlugin

class ChannelListener(private val events: EventChannel.EventSink) : TwilioChannelListener {
    override fun onMessageAdded(message: Message) {
        TwilioProgrammableChatPlugin.debug("ChannelListener.onMessageAdded => messageSid = ${message.sid}")
        sendEvent("messageAdded", mapOf("message" to Mapper.messageToMap(message)))
    }

    override fun onMessageUpdated(message: Message, reason: Message.UpdateReason) {
        TwilioProgrammableChatPlugin.debug("ChannelListener.onMessageUpdated => messageSid = ${message.sid}, reason = $reason")
        sendEvent("messageUpdated", mapOf(
                "message" to Mapper.messageToMap(message),
                "reason" to mapOf(
                        "type" to "message",
                        "value" to reason.toString()
                )
        ))
    }

    override fun onMessageDeleted(message: Message) {
        TwilioProgrammableChatPlugin.debug("ChannelListener.onMessageDeleted => messageSid = ${message.sid}")
        sendEvent("messageDeleted", mapOf("message" to Mapper.messageToMap(message)))
    }

    override fun onMemberAdded(member: Member) {
        TwilioProgrammableChatPlugin.debug("ChannelListener.onMemberAdded => memberSid = ${member.sid}")
        sendEvent("memberAdded", mapOf("member" to Mapper.memberToMap(member)))
    }

    override fun onMemberUpdated(member: Member, reason: Member.UpdateReason) {
        TwilioProgrammableChatPlugin.debug("ChannelListener.onMemberUpdated => memberSid = ${member.sid}, reason = $reason")
        sendEvent("memberUpdated", mapOf(
                "member" to Mapper.memberToMap(member),
                "reason" to mapOf(
                        "type" to "member",
                        "value" to reason.toString()
                )
        ))
    }

    override fun onMemberDeleted(member: Member) {
        TwilioProgrammableChatPlugin.debug("ChannelListener.onMemberDeleted => memberSid = ${member.sid}")
        sendEvent("memberDeleted", mapOf("member" to Mapper.memberToMap(member)))
    }

    override fun onTypingStarted(channel: Channel, member: Member) {
        TwilioProgrammableChatPlugin.debug("ChannelListener.onTypingStarted => channelSid = ${channel.sid}, memberSid = ${member.sid}")
        sendEvent("typingStarted", mapOf("channel" to Mapper.channelToMap(channel), "member" to Mapper.memberToMap(member)))
    }

    override fun onTypingEnded(channel: Channel, member: Member) {
        TwilioProgrammableChatPlugin.debug("ChannelListener.onTypingEnded => channelSid = ${channel.sid}, memberSid = ${member.sid}")
        sendEvent("typingEnded", mapOf("channel" to Mapper.channelToMap(channel), "member" to Mapper.memberToMap(member)))
    }

    override fun onSynchronizationChanged(channel: Channel) {
        TwilioProgrammableChatPlugin.debug("ChannelListener.onSynchronizationChanged => channelSid = ${channel.sid}")
        sendEvent("synchronizationChanged", mapOf("channel" to Mapper.channelToMap(channel)))
    }

    private fun sendEvent(name: String, data: Any?, e: ErrorInfo? = null) {
        val eventData = mapOf("name" to name, "data" to data, "error" to Mapper.errorInfoToMap(e))
        events.success(eventData)
    }
}

File: ios/Classes/Listeners/ChannelListener.swift
import Flutter
import TwilioChatClient

public class ChannelListener: NSObject, TCHChannelDelegate {
    let events: FlutterEventSink

    init(_ events: @escaping FlutterEventSink) {
        self.events = events
    }

    // onMessageAdded
    public func chatClient(_ client: TwilioChatClient, channel: TCHChannel, messageAdded message: TCHMessage) {
        SwiftTwilioProgrammableChatPlugin.debug(
            "ChannelListener.onMessageAdded => messageSid = \(String(describing: message.sid))")
        sendEvent("messageAdded", data: [
            "message": Mapper.messageToDict(message, channelSid: channel.sid)
        ])
    }

    // onMessageUpdated
    public func chatClient(
        _ client: TwilioChatClient, channel: TCHChannel, message: TCHMessage, updated: TCHMessageUpdate) {
        SwiftTwilioProgrammableChatPlugin.debug(
            "ChannelListener.onMessageUpdated => messageSid = \(String(describing: message.sid)), " +
            "updated = \(String(describing: updated))")
        sendEvent("messageUpdated", data: [
            "message": Mapper.messageToDict(message, channelSid: channel.sid),
            "reason": [
                "type": "message",
                "value": Mapper.messageUpdateToString(updated)
            ]
        ])
    }

    // onMessageDeleted
    public func chatClient(_ client: TwilioChatClient, channel: TCHChannel, messageDeleted message: TCHMessage) {
        SwiftTwilioProgrammableChatPlugin.debug(
            "ChannelListener.onMessageDeleted => messageSid = \(String(describing: message.sid))")
        sendEvent("messageDeleted", data: [
            "message": Mapper.messageToDict(message, channelSid: channel.sid)
        ])
    }

    // onMemberAdded
    public func chatClient(_ client: TwilioChatClient, channel: TCHChannel, memberJoined member: TCHMember) {
        SwiftTwilioProgrammableChatPlugin.debug(
            "ChannelListener.onMemberAdded => memberSid = \(String(describing: member.sid))")
        sendEvent("memberAdded", data: [
            "member": Mapper.memberToDict(member, channelSid: channel.sid) as Any
        ])
    }

    // onMemberUpdated
    public func chatClient(_ client: TwilioChatClient, channel: TCHChannel,
                           member: TCHMember, updated: TCHMemberUpdate) {
        SwiftTwilioProgrammableChatPlugin.debug(
            "ChannelListener.onMemberUpdated => memberSid = \(String(describing: member.sid)), " +
            "updated = \(String(describing: updated))")
        sendEvent("memberUpdated", data: [
            "member": Mapper.memberToDict(member, channelSid: channel.sid) as Any,
            "reason": [
                "type": "member",
                "value": Mapper.memberUpdateToString(updated)
            ]
        ])
    }

    // onMemberDeleted
    public func chatClient(_ client: TwilioChatClient, channel: TCHChannel, memberLeft member: TCHMember) {
        SwiftTwilioProgrammableChatPlugin.debug(
            "ChannelListener.onMemberDeleted => memberSid = \(String(describing: member.sid))")
        sendEvent("memberDeleted", data: [
            "member": Mapper.memberToDict(member, channelSid: channel.sid) as Any
        ])
    }

    // onTypingStarted
    public func chatClient(_ client: TwilioChatClient, typingStartedOn channel: TCHChannel, member: TCHMember) {
        SwiftTwilioProgrammableChatPlugin.debug(
            "ChannelListener.onTypingStarted => channelSid = \(String(describing: channel.sid)), " +
            "memberSid = \(String(describing: member.sid))")
        sendEvent("typingStarted", data: [
            "channel": Mapper.channelToDict(channel) as Any,
            "member": Mapper.memberToDict(member, channelSid: channel.sid) as Any
        ])
    }

    // onTypingEnded
    public func chatClient(_ client: TwilioChatClient, typingEndedOn channel: TCHChannel, member: TCHMember) {
        SwiftTwilioProgrammableChatPlugin.debug(
            "ChannelListener.onTypingEnded => channelSid = \(String(describing: channel.sid)), " +
            "memberSid = \(String(describing: member.sid))")
        sendEvent("typingEnded", data: [
            "channel": Mapper.channelToDict(channel) as Any,
            "member": Mapper.memberToDict(member, channelSid: channel.sid) as Any
        ])
    }

    // onSynchronizationChanged
    public func chatClient(_ client: TwilioChatClient, channel: TCHChannel,
                           synchronizationStatusUpdated status: TCHChannelSynchronizationStatus) {
        SwiftTwilioProgrammableChatPlugin.debug(
            "ChannelListener.onSynchronizationChanged => channelSid = \(String(describing: channel.sid))")
        sendEvent("synchronizationChanged", data: [
            "channel": Mapper.channelToDict(channel) as Any
        ])
    }

    private func sendEvent(_ name: String, data: [String: Any]? = nil, error: Error? = nil) {
        let eventData = [
            "name": name,
            "data": data,
            "error": Mapper.errorToDict(error)
            ] as [String: Any?]

        events(eventData)
    }
}
